vyper/semantics/analysis/local.py

```diff
@@ -914,14 +914,14 @@ def visit_Subscript(self, node: vy_ast.Subscript, typ: VyperType) -> None:
         else:
             base_type = get_exact_type_from_node(node.value)
 
-        # get the correct type for the index, it might
-        # not be exactly base_type.key_type
-        # note: index_type is validated in types_from_Subscript
-        index_types = get_possible_types_from_node(node.slice)
-        index_type = index_types.pop()
+        if isinstance(base_type, HashMapT):
+            index_type = base_type.key_type
+        else:
+            # Arrays allow most int types as index: Take the least specific
```

Comment:
maybe it's worth refactoring into a function. `base_type.index_type_from_node(node.slice)`

Reply:
that generates more import cycles 
![image](https://github.com/vyperlang/vyper/assets/2369243/f307683e-6182-4af6-b3c6-1f88deb3ca30)